nifi-toolkit/nifi-toolkit-encrypt-config/src/main/groovy/org/apache/nifi/properties/ConfigEncryptionTool.groovy

```diff
@@ -0,0 +1,540 @@
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import javax.crypto.Cipher
+import java.nio.charset.StandardCharsets
+import java.security.KeyException
+import java.security.Security
+
+class ConfigEncryptionTool {
+    private static final Logger logger = LoggerFactory.getLogger(ConfigEncryptionTool.class)
+
+    public String bootstrapConfPath
+    public String niFiPropertiesPath
+    public String outputNiFiPropertiesPath
+    public String loginIdentityProvidersPath
+
+    private String keyHex
+    private String password
+    private NiFiProperties niFiProperties
+
+    private boolean usingPassword = true
+
+    private static final String HELP_ARG = "help"
+    private static final String BOOTSTRAP_CONF_ARG = "bootstrapConf"
+    private static final String NIFI_PROPERTIES_ARG = "niFiProperties"
+    private static final String OUTPUT_NIFI_PROPERTIES_ARG = "outputNiFiProperties"
+    private static final String KEY_ARG = "key"
+    private static final String PASSWORD_ARG = "password"
+    private static final String USE_KEY_ARG = "useRawKey"
+
+    private static final int MIN_PASSWORD_LENGTH = 12
+
+    // Strong parameters as of 12 Aug 2016
+    private static final int SCRYPT_N = 2**16
+    private static final int SCRYPT_R = 8
+    private static final int SCRYPT_P = 1
+
+    private static
+    final String BOOTSTRAP_KEY_COMMENT = "# Master key in hexadecimal format for encrypted sensitive configuration values"
+    private static final String BOOTSTRAP_KEY_PREFIX = "nifi.bootstrap.sensitive.key="
+    private static final String JAVA_HOME = "JAVA_HOME"
+    private static final String NIFI_TOOLKIT_HOME = "NIFI_TOOLKIT_HOME"
+    private static final String SEP = System.lineSeparator()
+
+    private static final String FOOTER = buildFooter()
+
+    private static
+    final String DEFAULT_DESCRIPTION = "This tool reads from a nifi.properties file with plain sensitive configuration values, prompts the user for a master key, and encrypts each value. It will replace the plain value with the protected value in the same file (or write to a new nifi.properties file if specified)."
+
+    private static String buildHeader(String description = DEFAULT_DESCRIPTION) {
+        "${SEP}${description}${SEP * 2}"
+    }
+
+    private static String buildFooter() {
+        "${SEP}Java home: ${System.getenv(JAVA_HOME)}${SEP}NiFi Toolkit home: ${System.getenv(NIFI_TOOLKIT_HOME)}"
+    }
+
+    private final Options options;
+    private final String header;
+
+
+    public ConfigEncryptionTool() {
+        this(DEFAULT_DESCRIPTION)
+    }
+
+    public ConfigEncryptionTool(String description) {
+        this.header = buildHeader(description)
+        this.options = new Options()
+        options.addOption("h", HELP_ARG, false, "Prints this usage message")
+        options.addOption("n", NIFI_PROPERTIES_ARG, true, "The nifi.properties file containing unprotected config values (will be overwritten)")
```

Comment:
The first time I read the usage when I saw that -n said "will be overwritten", I assumed that if I didn't specify -o then it would just overwrite the file specified with -n, but it actually wrote to nifi-toolkit-1.0.0-SNAPSHOT/conf/nifi.properies, maybe just some additional wording?